proposals/0387-bls-pubkey-management-in-vote-account.md

```diff
@@ -0,0 +1,191 @@
+Alpenglow launches.
+
+## Dependencies
+
+- Alpenglow is specified in [SIMD 326](https://github.com/solana-foundation/solana-improvement-documents/pull/326)
+
+- VoteStateV4 is specified in [SIMD 185](https://github.com/solana-foundation/solana-improvement-documents/pull/185)
+and it adds an optional BLS public key field
+
+- Requiring BLS public key for Alpenglow is specified in [SIMD 357](https://github.com/solana-foundation/solana-improvement-documents/pull/357)
+
+- The BLS verification syscall is specified in [SIMD 388](https://github.com/solana-foundation/solana-improvement-documents/pull/388)
+
+## New Terminology
+
+- **BLS public key**: The public key used in BLS signatures. It will be
+stored as a 48 bytes vector in vote account.
+
+## Detailed Design
+
+BLS keypairs can be generated randomly like ed25519 keypairs. But to save the
+users some trouble on keypair management, we chose to derive their BLS keypair
+used in Alpenglow votes based on their ed25519 vote keypair. In other words,
+with an existing ed25519 vote keypair, the users can safely regenerate the
+associated BLS keypair on demand.
+
+### User Operations Support
+
+Since a user’s BLS keypair always changes with the vote authority keypair, we
+list three different scenarios below.
+
+1. Creating a new vote account
+
+After VoteStateV4 (SIMD 185) is enabled everywhere, when users create a new
+vote account, the correct BLS public key will be automatically added to the
+new vote account. The BLS public key is derived from the vote authority keypair
+(if authorized voter is not specified, identity keypair is used).
+
+2. Changing vote authority keypair on existing vote account
+
+After VoteStateV4 (SIMD 185) is enabled everywhere, when users use
+vote-authorize-voter-checked to update vote authority keypair, the correct BLS
+public key will be automatically updated in the vote account. The BLS public
+key rotation happens at the same time when a new vote authority switch happens.
+
+3. Updating missing BLS public key on existing vote account
+
+This is a temporary work-around before all existing staked vote accounts have
+proper BLS public keys specified. We will add update-bls-pubkey command so it
+will use the given vote authority keypair to generate and update the BLS public
+key in the vote account. When everyone has a proper BLS public key in their
+vote accounts this command can be removed.
+
+### New instructions to vote programs
+
+In the following instructions, there is a proof_of_possession field associated
+with the new BLS public key. Since we need to verify the caller really holds
+the BLS public key claimed, the caller needs to prove possession of the BLS
+keypair to the vote program by signing the BLS public key with the
+corresponding BLS keypair. Then the vote program can do a syscall to actually
+verify the validity of BLS public key. If the verification fails, the
+transaction will fail.
```

Comment:
There's no precedent for native programs making syscalls that I know of. We should probably add a BLS program which can verify the BLS key proof of possession. Then, the vote program can invoke the BLS program to verify the proof.

Reply:
Added that to "Alternatives Considered".